pkg/controlplane/peer/certs.go

```diff
@@ -0,0 +1,165 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package peer
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/clusterlink-net/clusterlink/pkg/util/tls"
+)
+
+// CertsConsumer represents a consumer of peer TLS certificates.
+type CertsConsumer interface {
+	SetPeerCertificates(parsedCertData *tls.ParsedCertData, rawCertData *tls.RawCertData) error
+}
+
+// parseError represents an error in parsing the peer certificates.
+type parseError struct {
+	err error
+}
+
+func (e parseError) Error() string {
+	return e.err.Error()
+}
+
+// CertsWatcher watches certificate updates.
+type CertsWatcher struct {
+	caPath   string
+	certPath string
+	keyPath  string
+
+	stopCh    chan struct{}
+	consumers []CertsConsumer
+
+	logger *logrus.Entry
+}
+
+// Name of the watcher.
+func (w *CertsWatcher) Name() string {
+	return "certs-watcher"
+}
+
+// AddConsumer adds a new peer certificates consumer.
+func (w *CertsWatcher) AddConsumer(consumer CertsConsumer) {
+	w.consumers = append(w.consumers, consumer)
+}
+
+// ReadCertsAndUpdateConsumers reads the peer certificates and updates the consumers.
+func (w *CertsWatcher) ReadCertsAndUpdateConsumers() error {
+	w.logger.Infof("Updating certificates.")
+
+	parsedCertData, rawCertData, err := tls.ParseFiles(w.caPath, w.certPath, w.keyPath)
+	if err != nil {
+		return &parseError{err: err}
+	}
+
+	for _, consumer := range w.consumers {
+		if err := consumer.SetPeerCertificates(parsedCertData, rawCertData); err != nil {
+			return fmt.Errorf("error setting peer certificates on %v: %w", consumer, err)
+		}
+	}
+
+	return nil
+}
+
+// Start the certs watcher.
+func (w *CertsWatcher) Start() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("cannot initialize file watcher: %w", err)
+	}
+
+	defer func() {
+		if err := watcher.Close(); err != nil {
+			w.logger.Warnf("Cannot close watcher: %v", err)
+		}
+	}()
+
+	watchedFiles := []string{w.caPath, w.certPath, w.keyPath}
+	watchedDirs := make(map[string]interface{})
+	for _, file := range watchedFiles {
+		dir := path.Dir(file)
+		if _, ok := watchedDirs[dir]; !ok {
+			w.logger.Infof("Watching: %s.", dir)
+			if err := watcher.Add(dir); err != nil {
+				return fmt.Errorf("cannot watch directory '%s': %w", dir, err)
+			}
+			watchedDirs[dir] = nil
+		}
+	}
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	certsModified := false
+	for {
+		select {
+		case <-w.stopCh:
+			return nil
+		case event := <-watcher.Events:
+			w.logger.Infof("Event: %v", event)
+			certsModified = true
+		case err := <-watcher.Errors:
+			w.logger.Infof("Error: %v", err)
+			return err
+		case <-ticker.C:
```

Comment:
nit: an alternative to reloading on timer is to look at how configmap changes are detected and the symlinks followed to determine change. See for example in [viper](https://github.com/spf13/viper/commit/e0f7631cf3ac7e7530949c7e154855076b0a4c17).
I'm ok with it as is, but it may be a useful package in which case you may want to have that as a more complete implementation, not relying on timers.

Reply:
Sorry, I don't follow.
Let's talk about it in our next meeting.

Reply:
currently: mark the file change event, every second (on timer) reload if there was a change event.
viper: on change event - follow symlinks to detect if the actual file content changed and reload it.

Viper does this for configmap mounted as a volume which is similar to how secrets are mounted.
I was just saying that you *could* look at the code that actually examines the file that holds the data instead of reloading based on timer. It is not critical